main/src/main/scala/sbt/Defaults.scala

```diff
@@ -2356,17 +2457,75 @@ object Defaults extends BuildCommon {
     }
     analysis
   }
-  def compileIncrementalTask = Def.task {
-    val s = streams.value
-    val ci = (compile / compileInputs).value
-    val ping = earlyOutputPing.value
-    val reporter = (compile / bspReporter).value
-    BspCompileTask.compute(bspTargetIdentifier.value, thisProjectRef.value, configuration.value) {
-      task =>
-        // TODO - Should readAnalysis + saveAnalysis be scoped by the compile task too?
-        compileIncrementalTaskImpl(task, s, ci, ping, reporter)
-    }
-  }
+
+  def compileIncrementalTaskSettings(key: TaskKey[(Boolean, HashedVirtualFileRef)]) =
+    inTask(key)(
+      Seq(
+        (TaskZero / key) := (Def
+          .cachedTask {
+            val s = streams.value
+            val ci = (compile / compileInputs).value
+            // This is a cacheable version
+            val ci2 = (compile / compileInputs2).value
+            val ping = (TaskZero / earlyOutputPing).value
+            val reporter = (compile / bspReporter).value
+            val setup: Setup = (TaskZero / compileIncSetup).value
+            val useBinary: Boolean = enableBinaryCompileAnalysis.value
+            val c = fileConverter.value
+            val analysisResult: CompileResult =
+              BspCompileTask
+                .compute(bspTargetIdentifier.value, thisProjectRef.value, configuration.value) {
+                  bspTask =>
+                    // TODO - Should readAnalysis + saveAnalysis be scoped by the compile task too?
+                    compileIncrementalTaskImpl(bspTask, s, ci, ping, reporter)
+                }
```

Comment:
In a monorepo build tool, like Bazel, a cached compilation is treated like our Coursier library dependency JAR download, so the downstream targets (subprojects) would work, but the intermediate side effects like warnings aren't displayed anywhere. This kind of make sense because with Bazel you handle the entire company's code, you really don't want to be bombarded with warnings. We might revisit this, but initially I'm sort of inlined to display warning only if we perform actual compilation.

Reply:
I see. It is understandable that a cached task does not produce all the side effects (the logs, the BSP notifications) etc. 

In this case though I wonder if we should inform the user about it: `Compilation of foo restored from the cache`.

Also sometimes we might want to bypass the cache and force a compilation. It seems that `clean; compile` is not enough anymore. So how can we do so?

Reply:
We could probably provide some flag or setting to ignore cache?